src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,201 @@
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 declare(strict_types=1);
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string */
+	private static $serviceURL;
+
+	/**
+	 * Search the device node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $deviceType
+	 * @param \SimpleXMLElement $foundDevice reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function getChildDevice(\SimpleXMLElement $device, string $deviceType, &$foundDevice) : bool{
+		foreach($device->deviceList->device as $child){
+			if((string)$child->deviceType === $deviceType){
+				$foundDevice = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Search the service node with given type.
+	 *
+	 * @param \SimpleXMLElement $device
+	 * @param string $serviceType
+	 * @param \SimpleXMLElement $foundService reference parameter
+	 *
+	 * @return bool
+	 */
+	private static function getChildService(\SimpleXMLElement $device, string $serviceType, &$foundService) : bool{
+		foreach($device->serviceList->service as $child){
+			if((string)$child->serviceType === $serviceType){
+				$foundService = $child;
+				return true;
+			}
+		}
+		return false;
+	}
+	
+	public static function setup() : void{
+		$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+		socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, array("sec"=>3,"usec"=>0));
+		$contents =
+			"M-SEARCH * HTTP/1.1\r\n".
+			"MX: 2\r\n".
+			"HOST: 239.255.255.250:1900\r\n".
+			"MAN: \"ssdp:discover\"\r\n".
+			"ST: upnp:rootdevice\r\n\r\n";
+		socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
```

Comment:
This might emit E_WARNING, which will turn into `ErrorException` that will crash the server.